File: models/store/StockRequest.js
const { DataTypes } = require('sequelize');
const sequelize = require('../../config/database');

const StockRequest = sequelize.define('StockRequest', {
    id: {
        type: DataTypes.UUID,
        primaryKey: true,
        defaultValue: DataTypes.UUIDV4
    },
    institution_id: {
        type: DataTypes.UUID,
        allowNull: true
    },
    request_number: {
        type: DataTypes.STRING(50),
        unique: true,
        allowNull: false
    },
    requested_by: {
        type: DataTypes.UUID,
        allowNull: true
    },
    department_id: {
        type: DataTypes.UUID,
        allowNull: true
    },
    priority: {
        type: DataTypes.ENUM('low', 'medium', 'high', 'critical'),
        defaultValue: 'medium'
    },
    status: {
        type: DataTypes.ENUM('pending', 'approved', 'partially_approved', 'rejected', 'issued', 'cancelled'),
        defaultValue: 'pending'
    },
    request_date: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW
    },
    required_date: {
        type: DataTypes.DATE,
        allowNull: true
    },
    purpose: {
        type: DataTypes.TEXT,
        allowNull: true
    },
    notes: {
        type: DataTypes.TEXT,
        allowNull: true
    },
    approved_by: {
        type: DataTypes.UUID,
        allowNull: true
    },
    approved_date: {
        type: DataTypes.DATE,
        allowNull: true
    },
    rejection_reason: {
        type: DataTypes.TEXT,
        allowNull: true
    }
}, {
    tableName: 'stock_requests',
    timestamps: true,
});

// Define associations
StockRequest.associate = (models) => {
    StockRequest.hasMany(models.StockRequestItem, { foreignKey: 'stock_request_id', as: 'items' });
    StockRequest.belongsTo(models.Item, { foreignKey: 'item_id', as: 'item' });
};

module.exports = StockRequest;

File: store-service/app/models/stock_request.py
"""StockRequest model - department stock requests."""

import uuid
from datetime import datetime

from sqlalchemy import Column, DateTime, Enum, String, Text
from sqlalchemy.dialects.postgresql import UUID

from app.database import Base
from app.utils.types import RequestStatus, Priority


class StockRequest(Base):
    __tablename__ = "stock_requests"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    institution_id = Column(UUID(as_uuid=True), nullable=True)
    request_number = Column(String(50), unique=True, nullable=False)
    requested_by = Column(UUID(as_uuid=True), nullable=True)
    department_id = Column(UUID(as_uuid=True), nullable=True)
    priority = Column(Enum(Priority), default=Priority.MEDIUM)
    status = Column(Enum(RequestStatus), default=RequestStatus.PENDING)
    request_date = Column(DateTime(timezone=True), default=datetime.utcnow)
    required_date = Column(DateTime(timezone=True), nullable=True)
    purpose = Column(Text, nullable=True)
    notes = Column(Text, nullable=True)
    approved_by = Column(UUID(as_uuid=True), nullable=True)
    approved_date = Column(DateTime(timezone=True), nullable=True)
    rejection_reason = Column(Text, nullable=True)

    created_at = Column("createdAt", DateTime(timezone=True), default=datetime.utcnow)
    updated_at = Column("updatedAt", DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow)
